File: files/root/www/luci-static/resources/view/fzu-network.js
'use strict';
'require view';
'require form';
'require rpc';
'require ui';
'require poll';

var callStatus = rpc.declare({
	object: 'luci.fzu-network',
	method: 'status',
	expect: { online: false, msg: '', user: '-', ip: '-', mac: '-', log: '' }
});

var callClearLog = rpc.declare({
	object: 'luci.fzu-network',
	method: 'clearlog',
	expect: {}
});

function renderStatusTable(status) {
	return E('div', { 'class': 'cbi-section status-section' }, [
		E('h3', {}, _('状态')),
		E('div', { 'class': 'cbi-section-node' }, [
			E('table', { 'class': 'table' }, [
				E('tr', { 'class': 'tr' }, [
					E('td', { 'class': 'td left', 'style': 'width:33%' }, _('认证状态')),
					E('td', { 'class': 'td' }, status.online
						? E('span', { 'style': 'color:green;font-weight:bold' }, _('在线'))
						: E('div', {}, [
							E('span', { 'style': 'color:red;font-weight:bold' }, _('离线')),
							' ',
							E('em', {}, status.msg || '')
						])
					)
				]),
				E('tr', { 'class': 'tr' }, [
					E('td', { 'class': 'td left' }, _('登录账号')),
					E('td', { 'class': 'td' }, status.user || '-')
				]),
				E('tr', { 'class': 'tr' }, [
					E('td', { 'class': 'td left' }, _('IP 地址')),
					E('td', { 'class': 'td' }, status.ip || '-')
				]),
				E('tr', { 'class': 'tr' }, [
					E('td', { 'class': 'td left' }, _('MAC 地址')),
					E('td', { 'class': 'td' }, status.mac || '-')
				])
			])
		])
	]);
}

function renderLog(status) {
	return E('div', { 'class': 'cbi-section' }, [
		E('h3', {}, _('日志')),
		E('div', { 'class': 'cbi-section-node' }, [
			E('pre', {
				'id': 'fzu-log',
				'style': 'max-height:300px;overflow:auto;font-size:12px;white-space:pre-wrap;word-break:break-all'
			}, status.log || _('（暂无日志）')),
			E('div', { 'style': 'margin-top:8px' }, [
				E('button', {
					'class': 'btn cbi-button cbi-button-action',
					'click': ui.createHandlerFn(null, function() {
						return callClearLog().then(function() {
							var pre = document.getElementById('fzu-log');
							if (pre) pre.textContent = _('（暂无日志）');

						});
					})
				}, _('清除日志'))
			])
		])
	]);
}

return view.extend({
	load: function() {
		return callStatus();
	},

	render: function(status) {
		var m, s, o;

		m = new form.Map('fzu-network', '', '');

		s = m.section(form.NamedSection, 'base', 'base', _('设置'));
		s.anonymous = true;

		o = s.option(form.Flag, 'enable', _('启用'));
		o.rmempty = false;

		o = s.option(form.Value, 'school_no', _('学号'));
		o.rmempty = false;

		o = s.option(form.Value, 'password', _('密码'));
		o.password = true;
		o.rmempty = false;

		return m.render().then(function(formNode) {
			var statusNode = renderStatusTable(status);
			var node = E('div', {}, [
				statusNode,
				formNode,
				renderLog(status)
			]);

			poll.add(function() {
				return callStatus().then(function(st) {
					var newStatus = renderStatusTable(st);
					statusNode.parentNode.replaceChild(newStatus, statusNode);
					statusNode = newStatus;
					var pre = node.querySelector('#fzu-log');
					if (pre) pre.textContent = st.log || _('（暂无日志）');
				});
			}, 5);

			return node;
		});
	},

	handleSaveApply: null,
	handleSave: null,
	handleReset: null
});

File: files/root/usr/lib/lua/luci/model/cbi/fzu-network.lua
local fs = require("nixio.fs")
require("luci.sys")

local m, s

-- 查询校园网在线状态
local status_json = luci.sys.exec(
	"curl -s -X POST -H 'Content-Type: application/x-www-form-urlencoded; charset=UTF-8' " ..
	"--data-urlencode 'userIndex=' " ..
	"'http://172.16.0.46/eportal/InterFace.do?method=getOnlineUserInfo' 2>/dev/null"
)

local online = false
local info_result, info_msg, info_user, info_ip, info_mac = "未知", "无法连接到认证服务器", "-", "-", "-"

if status_json and #status_json > 0 then
	local ok, data = pcall(function()
		-- 简单解析 JSON 字段
		local function jget(str, key)
			return str:match('"' .. key .. '"%s*:%s*"([^"]*)"') or "-"
		end
		return {
			result  = jget(status_json, "result"),
			message = jget(status_json, "message"),
			user    = jget(status_json, "userName"),
			ip      = jget(status_json, "userIp"),
			mac     = jget(status_json, "userMac"),
		}
	end)
	if ok and data then
		info_result = data.result
		info_msg    = data.message
		info_user   = data.user
		info_ip     = data.ip
		info_mac    = data.mac
		online = (data.result == "success")
	end
end

m = Map("fzu-network", "福州大学校园网", "自动登录福州大学校园网认证。")

-- 在线状态展示
s = m:section(TypedSection, "base", "当前连接状态")
s.anonymous = true

local st = s:option(DummyValue, "_status", "认证状态")
st.rawhtml = true
function st.cfgvalue(self, section)
	if online then
		return '<span style="color:green;font-weight:bold;">● 已在线</span>'
	else
		return '<span style="color:red;font-weight:bold;">● 未在线</span> &nbsp; <em>' .. info_msg .. '</em>'
	end
end

local dv_user = s:option(DummyValue, "_user", "登录账号")
function dv_user.cfgvalue(self, section) return info_user end

local dv_ip = s:option(DummyValue, "_ip", "IP地址")
function dv_ip.cfgvalue(self, section) return info_ip end

local dv_mac = s:option(DummyValue, "_mac", "MAC地址")
function dv_mac.cfgvalue(self, section) return info_mac end

-- 基本设置
s = m:section(TypedSection, "base", "基本设置")
s.anonymous = true

enable = s:option(Flag, "enable", "启用")
enable.rmempty = false

school_no = s:option(Value, "school_no", "学号")
school_no.rmempty = false

password = s:option(Value, "password", "密码")
password.password = true
password.rmempty = false

time = s:option(Value, "time", "检测间隔", "Crontab表达式")
time.rmempty = false
time.default = "*/5 * * * *"

user_agent = s:option(Value, "user_agent", "User Agent")
user_agent.rmempty = false

local spacer = s:option(DummyValue, "_spacer", "")
spacer.rawhtml = true
function spacer.cfgvalue(self, section) return '<div style="margin-bottom:8px"></div>' end

-- 运行日志
s = m:section(TypedSection, "base", "运行日志")
s.anonymous = true

local log_file = "/var/log/fzu-network.log"
tvlog = s:option(TextValue, "sylogtext")
tvlog.rows = 16
tvlog.readonly = "readonly"
tvlog.wrap = "off"

function tvlog.cfgvalue(self, section)
	local content = ""
	if log_file and nixio.fs.access(log_file) then
		content = luci.sys.exec("tail -n 100 %s" % log_file)
	end
	return content
end

tvlog.write = function(self, section, value) end

function m.on_after_commit(map)
	luci.sys.call("/etc/init.d/fzu-network restart >/dev/null 2>&1 &")
end

-- 清空日志按钮 + 自动刷新状态的 JS
local refresh_js = s:option(DummyValue, "_refresh_js", "")
refresh_js.rawhtml = true
function refresh_js.cfgvalue(self, section)
	return [[<a href="javascript:void(0)" onclick="clearLog()" style="float:right;font-size:0.85em;">清空日志</a>
<script type="text/javascript">
function clearLog() {
	var xhr = new XMLHttpRequest();
	xhr.open('GET', L.url('admin/services/fzu-network/clearlog'), true);
	xhr.onreadystatechange = function() {
		if (xhr.readyState === 4) {
			var logEl = document.querySelector('textarea[id$="sylogtext"]');
			if (logEl) logEl.value = '';
		}
	};
	xhr.send();
}
</script>

<script type="text/javascript">
(function() {
	var interval = 5000;
	function updateStatus() {
		var xhr = new XMLHttpRequest();
		xhr.open('GET', L.url('admin/services/fzu-network/status'), true);
		xhr.onreadystatechange = function() {
			if (xhr.readyState === 4 && xhr.status === 200) {
				try {
					var d = JSON.parse(xhr.responseText)[0];
					var stEl = document.querySelector('[data-name="_status"] .cbi-value-field');
					if (stEl) stEl.innerHTML = d.online
						? '<span style="color:green;font-weight:bold;">● 已在线</span>'
						: '<span style="color:red;font-weight:bold;">● 未在线</span> &nbsp; <em>' + d.msg + '</em>';
					var userEl = document.querySelector('[data-name="_user"] .cbi-value-field');
					if (userEl) userEl.textContent = d.user;
					var ipEl = document.querySelector('[data-name="_ip"] .cbi-value-field');
					if (ipEl) ipEl.textContent = d.ip;
					var macEl = document.querySelector('[data-name="_mac"] .cbi-value-field');
					if (macEl) macEl.textContent = d.mac;
					var logEl = document.querySelector('textarea[id$="sylogtext"]');
					if (logEl) { logEl.value = d.log; logEl.scrollTop = logEl.scrollHeight; }
				} catch(e) {}
			}
		};
		xhr.send();
	}
	setTimeout(function tick() { updateStatus(); setTimeout(tick, interval); }, interval);
})();
</script>
]]
end

return m


